osu.Game/Online/Multiplayer/MultiplayerClient.cs

```diff
@@ -50,80 +52,68 @@ private void apiStateChanged(ValueChangedEvent<APIState> state)
             {
                 case APIState.Failing:
                 case APIState.Offline:
-                    connection?.StopAsync();
-                    connection = null;
+                    Task.Run(Disconnect).CatchUnobservedExceptions();
                     break;
 
                 case APIState.Online:
-                    Task.Run(Connect);
+                    Task.Run(Connect).CatchUnobservedExceptions();
                     break;
             }
         }
 
-        protected virtual async Task Connect()
-        {
-            if (connection != null)
-                return;
-
-            connection = new HubConnectionBuilder()
-                         .WithUrl(endpoint, options =>
-                         {
-                             options.Headers.Add("Authorization", $"Bearer {api.AccessToken}");
-                         })
-                         .AddNewtonsoftJsonProtocol(options => { options.PayloadSerializerSettings.ReferenceLoopHandling = ReferenceLoopHandling.Ignore; })
-                         .Build();
-
-            // this is kind of SILLY
-            // https://github.com/dotnet/aspnetcore/issues/15198
-            connection.On<MultiplayerRoomState>(nameof(IMultiplayerClient.RoomStateChanged), ((IMultiplayerClient)this).RoomStateChanged);
-            connection.On<MultiplayerRoomUser>(nameof(IMultiplayerClient.UserJoined), ((IMultiplayerClient)this).UserJoined);
-            connection.On<MultiplayerRoomUser>(nameof(IMultiplayerClient.UserLeft), ((IMultiplayerClient)this).UserLeft);
-            connection.On<int>(nameof(IMultiplayerClient.HostChanged), ((IMultiplayerClient)this).HostChanged);
-            connection.On<MultiplayerRoomSettings>(nameof(IMultiplayerClient.SettingsChanged), ((IMultiplayerClient)this).SettingsChanged);
-            connection.On<int, MultiplayerUserState>(nameof(IMultiplayerClient.UserStateChanged), ((IMultiplayerClient)this).UserStateChanged);
-            connection.On(nameof(IMultiplayerClient.LoadRequested), ((IMultiplayerClient)this).LoadRequested);
-            connection.On(nameof(IMultiplayerClient.MatchStarted), ((IMultiplayerClient)this).MatchStarted);
-            connection.On(nameof(IMultiplayerClient.ResultsReady), ((IMultiplayerClient)this).ResultsReady);
-
-            connection.Closed += async ex =>
-            {
-                isConnected.Value = false;
+        private readonly SemaphoreSlim connectionLock = new SemaphoreSlim(1);
```

Comment:
Given what we've learned of `SemaphoreSlim` [over at the spectator server](https://github.com/ppy/osu-server-spectator/pull/24#discussion_r550190030), it might be wiser to explicitly release and dispose this semaphore when the entire client is getting disposed.

What I'm particularly worried of is a deadlock in the "retry connecting" case, as that case is holding the lock potentially forever.

Reply:
I've made some changes around this feedback, but I don't think we need to release the semaphore itself, since it's only being used internally. It may be different if the threads we were running were non-background and potentially keeping the application running.

I *have* added a force disconnect on disposal, though.

Reply:
Yeah, looks like that should be enough to drop the lock.